src/System.Threading/tests/SpinLockTests.cs

```diff
@@ -0,0 +1,395 @@
+        /// a counter variable and failed threads increment failed variable, count + failed must be equal to n
+        /// </summary>
+        /// <param name="threadsCount">Number of threads that call enter/exit</param>
+        /// <returns>True if succeeded, false otherwise</returns>
+        private static void RunSpinLockTest1_TryEnter(int threadsCount, bool enableThreadIDs)
+        {
+            for (int j = 0; j < 2; j++)
+            {
+                bool useMemoryBarrier = j == 0;
+                Task[] threads = new Task[threadsCount];
+                SpinLock slock = new SpinLock(enableThreadIDs);
+                int succeeded = 0;
+                int failed = 0;
+
+
+                // Run threads
+                for (int i = 0; i < threadsCount; i++)
+                {
+                    threads[i] = new Task(delegate ()
+                    {
+                        bool lockTaken = false;
+                        slock.TryEnter(ref lockTaken);
+                        if (lockTaken)
+                        {
+                            // Increment succeeded counter 
+                            Interlocked.Increment(ref succeeded);
+                            slock.Exit(useMemoryBarrier);
+                        }
+                        else
+                        {
+                            // Increment failed counter
+                            Interlocked.Increment(ref failed);
+                        }
+                    });
+                    threads[i].Start();
+                }
+                // Wait all threads
+                for (int i = 0; i < threadsCount; i++)
+                {
+                    threads[i].Wait();
+                }
+                // succeeded + failed must be equal to the threads count.
+                if (succeeded + failed != threadsCount)
+                {
+                    Assert.True(false, string.Format("SpinLock.TryEnter() failed, actual count: " + (succeeded + failed) +
+                        " expected :" + threadsCount));
+                }
+            }
+        }
+
+        /// <summary>
+        /// Test SpinLock.TryEnter(Timespan) by generating random timespan milliseconds
+        /// </summary>
+        /// <param name="threadsCount">Number of threads that call enter/exit</param>
+        /// <returns>True if succeeded, false otherwise</returns>
+        private static void RunSpinLockTest2_TryEnter(int threadsCount, bool enableThreadIDs)
+        {
+            for (int j = 0; j < 2; j++)
+            {
+                bool useMemoryBarrier = j == 0;
+                Task[] threads = new Task[threadsCount];
+                SpinLock slock = new SpinLock(enableThreadIDs);
+                int succeeded = 0;
+                int failed = 0;
+
+                // Run threads
+                for (int i = 0; i < threadsCount; i++)
+                {
+                    threads[i] = new Task(delegate (object x)
```

Comment:
This one uses Action<object> and passes in "i", should I just use Task.Factory.StartNew, or can I just pass the TaskScheduler into task.Start()?